src/pybind/mgr/dashboard/controllers/nvmeof.py

```
@@ -400,7 +401,26 @@ def update(
             w_mbytes_per_second: Optional[int] = None,
             gw_group: Optional[str] = None
         ):
-            if rbd_image_size:
+            resize = bool(rbd_image_size) 
+            chg_lb_group = bool(load_balancing_group)
+            set_qos = bool(rw_ios_per_second or rw_mbytes_per_second or r_mbytes_per_second or w_mbytes_per_second)            
+            opeartion_count = [resize, chg_lb_group, set_qos].count(True)
+            if opeartion_count < 1:
+                raise DashboardException(
+                    msg="no operation requested, aborting.",
+                    code='no_update_operations',
+                    http_status_code=400,
+                    component="nvmeof",
+                )
+            if opeartion_count > 1:
+                raise DashboardException(
+                    msg="More than 1 operation requested, aborting.",
+                    code='multiple_update_operations',
+                    http_status_code=400,
+                    component="nvmeof",
+                )
```

why are we restricting the multiple operations? Is there anything that could be broken here due if we go and update multiple items at once?

---

Just thought that in the case of an error in 1 of 3 operations we can be stuck with a really hard situation - we ended up in a state which is not the original state but also not the desired state, and from here we have to find a way to recover or at least make clear what happened to the user - both are not trivial as I see it :)

---

I think I understand what you mean, but I have a few points in mind that make me wonder. 

1. The current state (at least implementation-wise) is a "false declarative" - the API seems declarative but it fails if you would use it in a declarative manner.

2. I am not sure I am right about it, but I also a bit worried that since there are default parameters in this route, its not really declarative. I mean that the user (as the dashboard currently do), might pass only `rbd_image_size` param, and then we will only act regarding to this param and resize the namespace. it is not clear from this request what is the 'lb group' or the 'qos', as I would expect from a declarative API isn't it?

3. Another point I am a bit worried about is that the logic of recovery as we discussed above might be well relevant in clients as the dashboard or vsphere plugin since we can control them, but in the case of the CLI, the actual human user at the end who should do the recovery by himself, which worries me a bit - can we expect a typical Ceph user to interpret this kind of  response which is complex and know how to recover correctly? 

point #3 makes me wonder - maybe the `update` route makes sense in the context of an REST API, but not in a CLI context? Maybe another approach might be leave the route as is (API-wise), but make separate CLI commands for each operation.

Together with the concerns I mentioned, I understand from both of you that we want it to be declarative. As I understand, to reach it the following steps should take place:
1. Decide if we prefer the fail-fast or best-effort approach
2. Decide how the response should look like in case of a failure in the update route 
3. Implement the namespace update route accordingly.
4. Go over the clients and update them accordingly - I will probably need your guidance here. also maybe the clients today don't rely on our response at all and it is not a blocker at the moment? 🤔 

I will be glad to hear your thoughts!

---

Don't worry for the vsphere, since it's been designed with this behaviour in mind, and also it modifies things one by one (e.g.: resize).

I agree with you that it's "false declarative", but in the real world purity is almost unachievable, so most times we deal with mixed imperative and declarative approaches (e.g.: "rm -rf dir/" should delete everything under `dir/` but Linux users are used to receive errors and assume it might not delete everything and in case of failure it won't "undelete" the already deleted files/dirs).

From the REST API, I think the 202 response is good enough. While Ceph ensures strong consistency on the data, some things around Ceph are rather eventually consistent.

On you 2nd point, I don't remember the default params or any issues with them. I assume that if unspecified, the current namespace values will persist, and only explicit params will change. Isn't it that way?

On 3rd point, if you mean the rollback process, you can simply print the human readable error message. It'll be uncommon that a user wanted to change both namespace size, LGB and QoS params, but they would have the option to do so if they want. For REST API consumers, they should always revalidate the state after operations, so the source of truth should come the REST API, not a local state store.

---

Thanks for your response! 

> On you 2nd point, I don't remember the default params or any issues with them. I assume that if unspecified, the current namespace values will persist, and only explicit params will change. Isn't it that way?

it is indeed this way. I just thought (at least this is what I am used to from tiny experience with Terraform language [which is declarative]), that when you miss a field in such route, it should update it to its default value, but I guess it makes sense the other way around as well. 
When thinking about again, even the fact that we have a separate routes for creation and update are contradicting the declarative approach (at least as I conceive it), isn't it?

btw, what is your preference? best effort or fail fast? Frankly, I don't have a preference of my own, and will be glad to hear your thoughts regarding what would fit us better.

---

Also another question that came to mind - Assuming I return 202 response with the latest namespace object  - how would the user know if the changes succeed or failed? even if everything failed I will return the namespace object as well right?

---

Yup. That's part of the API contract. This PUT operation is not guaranteed to be atomic. Hence, the API consumer should take the returned resource (namespace) as the source of truth, and act accordingly.

If the requested state was achieved, a 200 should be returned instead.